R/data_modify.R

```diff
@@ -276,8 +286,196 @@ data_modify.grouped_df <- function(data, ..., .if = NULL, .at = NULL, .modify =
+    # sanity check, this may happen when user wants to remove a variable
+    # e.g. data_modify(iris, Species = NULL)
+    if (is.null(symbol_string)) next
+    # extract string-expression, if we have any
+    if (startsWith(symbol_string, "as_expr")) {
+      symbol_string <- gsub("as_expr\\((.*)\\)", "\\1", symbol_string)
+    } else {
+      # no expression token found
+      symbol_string <- NULL
+    }
+    # here we found an expression token - convert string into a regular expression
+    if (!is.null(symbol_string)) {
+      # check if we have any symbols instead of strings as expression
+      symbol_string <- .evaluate_expression_in_string(symbol_string, data)
+      # remove quotes from strings and save symbol name
+      symbol_string <- .fix_quotes(symbol_string)
+      symbol_name <- names(dots)[i]
+      # convert string into language and replace in dots
+      return_value <- try(str2lang(symbol_string), silent = TRUE)
+      # sanity check - for invalid expressions, like
+      # data_modify(iris, a = as_expr(c("1 + 1", "2 + 2")))
+      # we get an error here
+      if (inherits(return_value, "try-error")) {
+        insight::format_error(paste0(
+          "Could not evaluate expression `", symbol_string[1], "`. ",
+          "Please check if it's correctly specified. If you think there's a bug ",
+          "in `data_modify()`, please file an issue at {.url https://github.com/easystats/datawizard/issues}"
+        ))
+      }
+      dots[[i]] <- return_value
+      names(dots)[i] <- symbol_name
+    }
+  }
+  dots
+}
+
+
 # helper -------------
 
+
+.evaluate_expression_in_string <- function(symbol_string, data) {
+  # check if we have any symbols instead of strings as expression, e.g.
+  # xpr <- "sepwid = 2 * Sepal.Width"
+  # data_modify(iris, as_expr(xpr))
+  #
+  # in this case, we need to evaluate the symbol (i.e. convert symbol string
+  # into a language expression and then evaluate)
+  symbol_string <- unlist(lapply(symbol_string, function(symbol_element) {
+    if (!startsWith(symbol_element, "\"")) {
+      return_value <- .dynEval(str2lang(symbol_element))
+      # dynEval might fail if we don't look in data - sanity check
+      if (identical(return_value, symbol_element)) {
+        return_value <- .dynEval(str2lang(symbol_element), data = data)
+      }
+      return_value
+    } else {
+      symbol_element
+    }
+  }), use.names = FALSE)
+  # now we should have the expression as character string. Next, we
+  # # remove quotes from strings
+  gsub("^\"(.*)\"$", "\\1", symbol_string)
+}
+
+
+.fix_quotes <- function(symbol_string) {
+  # if user uses double-quotes insinde double-quotes, these are escaped by
+  # "\", e.g. data_modify(iris, foo = as_expr("grepl(\"a\", Species)"))
+  # In this case, we have double-backslashes, which need to be removed.
+  # Furthermore, to avoid adding back backslashes, we replace by single quotes
+  gsub("\\", "", gsub("\"", "'", symbol_string, fixed = TRUE), fixed = TRUE)
```

Comment:
Do we need something more robust than this? What happens if one puts single quotes in the regex of the `grepl()` example above?

Reply:
Then everything is fine. The problem is that double-quotes are escaped by a backslash, single quotes are not.

Reply:
I added some tests for all possible combinations